pkg/portal/cluster.go

```diff
@@ -190,3 +195,59 @@ func (p *portal) machineSets(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Type", "application/json")
 	_, _ = w.Write(b)
 }
+
+func (p *portal) statistics(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	duration := r.URL.Query().Get("duration")
+	endTimeString := r.URL.Query().Get("endtime")
+	endTime, err := time.Parse(time.RFC3339, endTimeString)
+	if err != nil {
+		p.internalServerError(w, err)
+		return
+	}
+	apiVars := mux.Vars(r)
+	subscription := apiVars["subscription"]
+	resourceGroup := apiVars["resourceGroup"]
+	clusterName := apiVars["clusterName"]
+	statisticsType := apiVars["statisticsType"]
+	resourceID := strings.ToLower(
+		fmt.Sprintf(
+			"/subscriptions/%s/resourceGroups/%s/providers/Microsoft.RedHatOpenShift/openShiftClusters/%s",
+			subscription, resourceGroup, clusterName))
+
+	promQuery, err := cluster.GetPromQuery(statisticsType)
+	if err != nil {
+		p.internalServerError(w, err)
+		return
+	}
+	prom := prometheus.New(p.log, p.dbOpenShiftClusters, p.dialer)
+	httpClient, err := prom.Cli(ctx, resourceID)
+	if err != nil {
+		p.internalServerError(w, err)
+		return
+	}
+
+	fetcher, err := p.makeFetcher(ctx, r)
+	if err != nil {
+		p.internalServerError(w, err)
+		return
+	}
+	promHost, promScheme := prom.GetPrometheusHostAndScheme()
+	prometheusURL := promScheme + "://" + promHost
+	APIStatistics, err := fetcher.Statistics(ctx, httpClient, promQuery, duration, endTime, prometheusURL)
+	if err != nil {
+		p.internalServerError(w, err)
+		return
+	}
```

Comment:
Query parameters aren't being validated until this function. This means that if a client provided invalid input data it would receive a `500 Internal Server Error` which would really explain what has actually happened. I'd expect to see input validation inside this handler so that you can return to the client a `400 Bad Request` in the case the client is sending something we don't expect.

Reply:
Yup you are right ... I have made the changes in the code and will pushed them shortly.